payment-service/app.py

```diff
@@ -1,19 +1,103 @@
-from flask import Flask
-from cassandra.cluster import Cluster
+import logging
+from http import HTTPStatus
 
-app = Flask(__name__)
+from flask import Flask, jsonify
+from databases.cassandra import CassandraDatabase
+from databases.postgres import PostgresDatabase
 
-# TODO: Dummy app for now. replace with real microservices
-@app.route('/')
-def hello_world():
-    return 'Hello, World!'
+LOGGER = logging.getLogger()
+LOGGER.setLevel('DEBUG')
+handler = logging.StreamHandler()
+handler.setFormatter(logging.Formatter(
+    "%(asctime)s [%(levelname)s] %(name)s: %(message)s"))
+LOGGER.addHandler(handler)
+app = Flask("payment-service")
 
 
-def cassandra_test():
-    cluster = Cluster(['127.0.0.1'])
-    session = cluster.connect()
+@app.route('/payment/pay/<uuid:user_id>/<uuid:order_id>/<float:amount>', methods=['POST'])
+def pay_order(user_id, order_id, amount):
+    LOGGER.info("Trying to pay order %s", order_id)
+    try:
+        success = database.subtract_credit(user_id, amount)
+        if success:
+            database.add_payment(order_id, 1, amount)
+            return HTTPStatus.OK
+        else:
+            database.add_payment(order_id, 0, amount)
+            return jsonify({'message': 'not enough credit'}), HTTPStatus.BAD_REQUEST
+    except RuntimeError:
+        return jsonify({'message': 'failure'}), HTTPStatus.BAD_REQUEST
+
+
+@app.route('/payment/cancel/<uuid:user_id>/<uuid:order_id>', methods=['POST'])
+def cancel_payment(user_id, order_id):
+    LOGGER.info("Canceling payment for %s by %s", (order_id, user_id))
+    try:
+        success_cancel, amount = database.cancel_payment(order_id)
+        if success_cancel:
+            success_add = database.add_credit(user_id, amount)
+            if success_add:
+                return HTTPStatus.OK
+            else:
+                return jsonify({'message': 'User not found'}), HTTPStatus.BAD_REQUEST
+        return jsonify({'message': 'Payment not found'}), HTTPStatus.BAD_REQUEST
+    except RuntimeError:
+        return jsonify({'message': 'failure'}), HTTPStatus.BAD_REQUEST
```

Comment:
BAD_REQUEST means client fault but RuntimeError is a server side error, therefore should return 500 (HTTPStatus.INTERNAL_SERVER_ERROR)

Reply:
The group project document specifies to keep the errors to only 2xx and 4xx, so that is why I didn't use any 500 codes.

Reply:
I guess it's not really needed, also if we get a server error it likely won't completely any more requests (db crashed or connection issues). This is more meant for completeness (going beyond what they expect) plus we have it in the other services so why not have it everywhere 😄 .

Reply:
True, I'll change it. Though I wonder if the reason they don't want us to use 5xx codes is that their evaluation method will not detect them, which would mean this could cost us points